planner/core/find_best_task.go

```diff
@@ -281,8 +289,54 @@ func (p *baseLogicalPlan) enumeratePhysicalPlans4Task(physicalPlans []PhysicalPl
 	return bestTask, cntPlan, nil
 }
 
+type physicalOptimizeOp struct {
+	// tracer is goring to track optimize steps during physical optimizing
+	tracer *tracing.PhysicalOptimizeTracer
+}
+
+func defaultPhysicalOptimizeOption() *physicalOptimizeOp {
+	return &physicalOptimizeOp{}
+}
+
+func (op *physicalOptimizeOp) withEnableOptimizeTracer(tracer *tracing.PhysicalOptimizeTracer) *physicalOptimizeOp {
+	op.tracer = tracer
+	return op
+}
+
+func (op *physicalOptimizeOp) appendChildToCurrent(candidateInfo *tracing.TaskInfo, plan Plan, prop string) *tracing.PhysicalOptimizeTraceInfo {
+	if op.tracer == nil || candidateInfo == nil {
+		return nil
+	}
+	info := buildPhysicalOptimizeTraceInfo(plan, prop)
+	return op.tracer.AppendPhysicalOptimizeToCurrent(candidateInfo, info)
+}
+
+func (op *physicalOptimizeOp) setCurrent(info *tracing.PhysicalOptimizeTraceInfo) {
+	if op.tracer == nil {
+		return
+	}
+	op.tracer.Current = info
+}
+
+func (op *physicalOptimizeOp) setBest(info *tracing.TaskInfo) {
+	if op.tracer == nil {
+		return
+	}
+	op.tracer.Current.BestTask = info
+}
+
+func (op *physicalOptimizeOp) appendCandidate(logicalPlan string, physicalPlan string) *tracing.TaskInfo {
+	if op.tracer == nil {
+		return nil
+	}
```

Comment:
I tend to pass `LogicalPlan` and PhsicalPlan here in order to skip running `fmt.Sprintf` if `op.tracer == nil`